### ISSUE TITLE
Prefer small-scope rule

### ISSUE THREAD

**Issue by BridgeAR** [eslint/eslint#11499]
**Please describe what the rule should do:**

The idea is almost identical to https://github.com/eslint/eslint/issues/5284 by @mysticatea.

An attempt was made in https://github.com/eslint/eslint/pull/9265 by @not-an-aardvark to implement it but there are a couple of cases that are difficult to identify safely.

Safe cases should be:

* If the variable is not reassigned at all afterwards, then it should be declared as `const` and should be safe to move inside.
* It should also be safe if the variable is not accessed before declaration in the inner scope and it is not referencing itself anywhere inside of the declaration.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Should warn:

```js
/* eslint prefer-smaller-scope */

let foo;        /*error 'foo' declaration can move into a block 'if (a) {...}'.*/
if (a) {
    foo = 100;
    bar(foo);
} else {
    bar(200);
}
```

Should not warn:

```
/* eslint prefer-smaller-scope */

let foo;
if (a) {
    foo = 100;
} else {
    foo = 200:
}
baz(foo);

let flag; // `flag` is only ever used in the function, but still needs to be declared outside it.

function doSomethingAfterFirstCall() {
    if (flag) foo();
    flag = true;
}

let bar; // `bar` is only ever used in the function, but still needs to be declared outside it.

function doSomethingElseAfterFirstCall() {
    bar = bar || true;
}

// Require should be an exception.
const util = require('util');

function foo() {
	util.inspect('Do not warn for require')
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a common case and not only an edge case special to a few use cases. Therefore it seems like a good idea to implement this in core.

**Are you willing to submit a pull request to implement this rule?**

Yes

**Comment by platinumazure** [comment]
:+1:. As a note, I think this is probably unsafe to autofix, but I would be in favor of having a rule that reports these issues.
